File: src/controllers/projectController.js
const projectService = require('../services/projectService');
const auditService = require('../services/auditService');

/**
 * POST /api/v1/projects
 */
const createProject = async (req, res, next) => {
  try {
    const project = await projectService.createProject(req.body, req.user._id);

    await auditService.logActivity({
      userId: req.user._id,
      action: 'PROJECT_CREATED',
      entity: 'Project',
      entityId: project._id,
      description: `Project created: "${project.name}"`,
      ipAddress: req.ip,
      userAgent: req.get('User-Agent'),
    });

    res.status(201).json({
      success: true,
      message: 'Project created successfully',
      data: { project },
    });
  } catch (error) {
    next(error);
  }
};

/**
 * GET /api/v1/projects
 */
const getProjects = async (req, res, next) => {
  try {
    const result = await projectService.getProjects(req.pagination, req.user._id, req.user.role);

    res.json({
      success: true,
      message: 'Projects retrieved successfully',
      data: result.data,
      pagination: result.pagination,
    });
  } catch (error) {
    next(error);
  }
};

/**
 * GET /api/v1/projects/:id
 */
const getProject = async (req, res, next) => {
  try {
    const project = await projectService.getProjectById(req.params.id);
    res.json({
      success: true,
      message: 'Project retrieved successfully',
      data: { project },
    });
  } catch (error) {
    next(error);
  }
};

/**
 * PUT /api/v1/projects/:id
 */
const updateProject = async (req, res, next) => {
  try {
    const project = await projectService.updateProject(req.params.id, req.body, req.user._id);

    await auditService.logActivity({
      userId: req.user._id,
      action: 'PROJECT_UPDATED',
      entity: 'Project',
      entityId: project._id,
      description: `Project updated: "${project.name}"`,
      ipAddress: req.ip,
      userAgent: req.get('User-Agent'),
    });

    res.json({
      success: true,
      message: 'Project updated successfully',
      data: { project },
    });
  } catch (error) {
    next(error);
  }
};

/**
 * DELETE /api/v1/projects/:id
 */
const deleteProject = async (req, res, next) => {
  try {
    const result = await projectService.deleteProject(req.params.id, req.user._id);

    await auditService.logActivity({
      userId: req.user._id,
      action: 'PROJECT_DELETED',
      entity: 'Project',
      entityId: req.params.id,
      description: result.message,
      ipAddress: req.ip,
      userAgent: req.get('User-Agent'),
    });

    res.json({
      success: true,
      message: result.message,
      data: null,
    });
  } catch (error) {
    next(error);
  }
};

/**
 * POST /api/v1/projects/:id/members
 */
const addMember = async (req, res, next) => {
  try {
    const { userId, role } = req.body;
    const project = await projectService.addMember(req.params.id, userId, role, req.user._id);

    await auditService.logActivity({
      userId: req.user._id,
      action: 'PROJECT_MEMBER_ADDED',
      entity: 'Project',
      entityId: project._id,
      description: `Member added to project`,
      ipAddress: req.ip,
      userAgent: req.get('User-Agent'),
    });

    res.json({
      success: true,
      message: 'Member added successfully',
      data: { project },
    });
  } catch (error) {
    next(error);
  }
};

/**
 * DELETE /api/v1/projects/:id/members/:userId
 */
const removeMember = async (req, res, next) => {
  try {
    const project = await projectService.removeMember(req.params.id, req.params.userId, req.user._id);

    await auditService.logActivity({
      userId: req.user._id,
      action: 'PROJECT_MEMBER_REMOVED',
      entity: 'Project',
      entityId: project._id,
      description: `Member removed from project`,
      ipAddress: req.ip,
      userAgent: req.get('User-Agent'),
    });

    res.json({
      success: true,
      message: 'Member removed successfully',
      data: { project },
    });
  } catch (error) {
    next(error);
  }
};

module.exports = {
  createProject,
  getProjects,
  getProject,
  updateProject,
  deleteProject,
  addMember,
  removeMember,
};

File: backend/src/main/java/com/taskflow/controller/ProjectController.java
package com.taskflow.controller;

import com.taskflow.dto.request.ProjectRequests;
import com.taskflow.dto.response.ServiceResults;
import com.taskflow.entity.Project;
import com.taskflow.entity.User;
import com.taskflow.service.ProjectService;
import com.taskflow.util.ApiResponse;
import com.taskflow.util.PaginationUtils;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/v1/projects")
public class ProjectController {

    private final ProjectService projectService;

    public ProjectController(ProjectService projectService) {
        this.projectService = projectService;
    }

    @GetMapping
    public ResponseEntity<ApiResponse<List<Project>>> getProjects(
            @RequestParam(required = false) Integer page,
            @RequestParam(required = false) Integer limit,
            @AuthenticationPrincipal User currentUser) {
        ServiceResults.PageResult<Project> result = projectService.getProjects(
                PaginationUtils.resolve(page, limit), currentUser);
        return ResponseEntity.ok(ApiResponse.ok("Projects retrieved successfully", result.data(),
                result.pagination()));
    }

    @PostMapping
    public ResponseEntity<ApiResponse<Map<String, Object>>> createProject(
            @Valid @RequestBody ProjectRequests.CreateProjectRequest request,
            @AuthenticationPrincipal User currentUser, HttpServletRequest http) {
        Project project = projectService.createProject(request, currentUser, http.getRemoteAddr(),
                http.getHeader("User-Agent"));
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(ApiResponse.ok("Project created successfully", Map.of("project", project)));
    }

    @GetMapping("/{id}")
    public ResponseEntity<ApiResponse<Map<String, Object>>> getProject(@PathVariable Long id) {
        return ResponseEntity.ok(ApiResponse.ok("Project retrieved successfully",
                Map.of("project", projectService.getProjectById(id))));
    }

    @PutMapping("/{id}")
    public ResponseEntity<ApiResponse<Map<String, Object>>> updateProject(
            @PathVariable Long id, @Valid @RequestBody ProjectRequests.UpdateProjectRequest request,
            @AuthenticationPrincipal User currentUser, HttpServletRequest http) {
        Project project = projectService.updateProject(id, request, currentUser, http.getRemoteAddr(),
                http.getHeader("User-Agent"));
        return ResponseEntity.ok(ApiResponse.ok("Project updated successfully", Map.of("project", project)));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<ApiResponse<Void>> deleteProject(@PathVariable Long id,
                                                           @AuthenticationPrincipal User currentUser,
                                                           HttpServletRequest http) {
        projectService.deleteProject(id, currentUser, http.getRemoteAddr(), http.getHeader("User-Agent"));
        return ResponseEntity.ok(ApiResponse.ok("Project deleted successfully", null));
    }

    @PostMapping("/{id}/members")
    public ResponseEntity<ApiResponse<Map<String, Object>>> addMember(
            @PathVariable Long id, @RequestBody ProjectRequests.AddMemberRequest request,
            @AuthenticationPrincipal User currentUser, HttpServletRequest http) {
        Project project = projectService.addMember(id, request.userId(), request.role(), currentUser,
                http.getRemoteAddr(), http.getHeader("User-Agent"));
        return ResponseEntity.ok(ApiResponse.ok("Member added successfully", Map.of("project", project)));
    }

    @DeleteMapping("/{id}/members/{userId}")
    public ResponseEntity<ApiResponse<Map<String, Object>>> removeMember(
            @PathVariable Long id, @PathVariable Long userId,
            @AuthenticationPrincipal User currentUser, HttpServletRequest http) {
        Project project = projectService.removeMember(id, userId, currentUser, http.getRemoteAddr(),
                http.getHeader("User-Agent"));
        return ResponseEntity.ok(ApiResponse.ok("Member removed successfully", Map.of("project", project)));
    }
}
